naga/src/back/msl/mod.rs

```diff
@@ -220,6 +220,40 @@ impl Default for Options {
     }
 }
 
+/// A mapping of vertex buffers and their attributes to shader
+/// locations.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+#[cfg_attr(feature = "deserialize", derive(serde::Deserialize))]
+pub struct AttributeMapping {
+    /// Shader location associated with this attribute
+    pub id: u32,
+    /// Offset in bytes from start of vertex buffer structure
+    pub offset: u32,
+    /// Format code to help us unpack the attribute into the type
+    /// used by the shader. Codes correspond to a 0-based index of
+    /// <https://gpuweb.github.io/gpuweb/#enumdef-gpuvertexformat>.
+    /// The conversion process is described by
+    /// <https://gpuweb.github.io/gpuweb/#vertex-processing>.
+    pub format: u32,
+}
+
+/// A description of a vertex buffer with all the information we
+/// need to address the attributes within it.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+#[cfg_attr(feature = "deserialize", derive(serde::Deserialize))]
+pub struct VertexBufferMapping {
+    /// Shader location associated with this buffer
+    pub id: u32,
+    /// Number of elements in this buffer
+    pub count: u32,
```

Comment:
The size of the vertex buffers won't be known when we generate the shader (at pipeline creation time).

We will have to pass the size of these buffers via another uniform buffer (constant buffer in metal).
We have this mechanism already in place. Here is some of the relevant code:

https://github.com/gfx-rs/wgpu/blob/8e15707631f1db5612824dea122ea8b6a4e9b271/naga/src/back/msl/writer.rs#L3240-L3259

Reply:
Thanks for this direction. I've updated the code to use the _buffer_sizes param.